task/script/0.1/script.yaml

```diff
@@ -0,0 +1,108 @@
+        debugging this tasks' code.
+      default: "false"
+  steps:
+    - name: "execute-script"
+      image: "$(params.image)"
+      env:
+        - name: "SCRIPT_SOURCE"
+          value: "$(params.script_source)"
+      script: |
+        #!/usr/bin/env bash
+        set -eo pipefail
+        
+        # debug
+        if [[ "$(params.trace)" == "true" ]]; then
+          set -x
+        fi
+        
+        # prepare input script
+        cat <<EOF >/var/myScript
+        ${SCRIPT_SOURCE}
+        EOF
+        mkdir -p /tekton/results/
+        cat /var/myScript
+        chmod +x /var/myScript
+        set +e
+        
+        # execute input script
+        # this line makes it that the tekton task receives live logs (stdout and stderr) and in addition it is logged
+        # as result. credit goes to
+        # https://stackoverflow.com/questions/692000/how-do-i-write-standard-error-to-a-file-while-using-tee-with-a-pipe
+        # https://stackoverflow.com/questions/36847897/in-bash-how-can-i-force-a-flush-of-an-incomplete-line-printed-to-the-terminal
+        /var/myScript 1> >(tee "$(results.stdout.path)") 2> >(tee "$(results.stderr.path)" >&2)
+        EXIT_CODE="$?"
+        
+        # handle result
+        set -e
+        echo -n "${EXIT_CODE}" >"$(results.exit_code.path)"
+        # sleep here as workaround for flaky error that frequently stdout and stderr are not flushed.
+        # At least some tekton result were (actually only stderr and/or stdout) missing!
+        # I tried running through script binary (but this seems to combine stderr and stdout,
+        # I want separate results for this!
+        # Furthermore I tried stdbuf, had not effect on flakiness.
+        # So the sleep fixes the flakiness 100% for me.
+        # Seems that the tee subprocesses run and terminate independently and on slow systems they may return after
+        # myScript returns.
+        i=0
+        while [ ${i} -lt 10 ]; do
+          set +e
+          if [[ -f "$(results.stdout.path)" && -f "$(results.stderr.path)" && -f "$(results.exit_code.path)" ]]; then
+            exit "${EXIT_CODE}"
+          fi
+          ((i++))
+          sleep 1
+        done
+        echo "ERROR: timeout awaiting '$(results.stdout.path)', '$(results.stderr.path)' 
+              and '$(results.exit_code.path)'"
+        exit 1
+  results:
+    - name: stdout
+      description: script standard output
+    - name: stderr
+      description: script error output
+    - name: exit_code
+      description: >
+        Script exit code, note however that the task will fail (with the exit code of the failed script) in case of
+        failure and then the result will not be accessible, because a failed tekton task has no results atm!
+  workspaces:
+    - name: "workspace-a"
+      optional: true
+    - name: "workspace-b"
+      optional: true
+    - name: "workspace-c"
+      optional: true
+    - name: "workspace-d"
+      optional: true
```

Comment:
instead of 4 workspaces can we have only 3? Reason: affinity-assitant runs by default and it won't allow Pipeline to bind more than 2 PVC's simultaneously in a Task. Probably 3 workspaces can be such that one can be used to bind pvc, one for configmap and one for secrets and all 3 can be optional. WDYT?

Reply:
Ok, I reduced it to three in the follow-up push.

In the long run, it would be good if tekton supports more namespaces, but luckily three is actually what we use productively atm!